### PR TITLE
Update and integrate TUIControl with better documentation.

### PR DESCRIPTION
**Pull request by avaidyam**
Not many new features, but tons more documentation, a cleaned up header, and a bit more goodies.

### CODE REVIEW

**Comment by avaidyam** [comment]
@jwilling Here's the `TUIControl` stuff on its own.

File: lib/UIKit/TUIControl.h [hunk 1]
```diff
@@ -17,67 +17,159 @@
...
 typedef enum TUIControlState : NSUInteger {
-	TUIControlStateNormal		= 0,
-	TUIControlStateHighlighted	= 1 << 0,
-	TUIControlStateDisabled		= 1 << 1,
-	TUIControlStateSelected		= 1 << 2,
-	TUIControlStateNotKey		= 1 << 11,
-	TUIControlStateApplication	= 0x00FF0000,
-	TUIControlStateReserved		= 0xFF000000
+	
+	// The normal, or default state of a control—that is,
```
Comment (jwilling) [inline lib/UIKit/TUIControl.h]: This would read better as

> The default state of a control. It is enabled, but not selected or highlighted.

File: lib/UIKit/TUIControl.h [hunk 2]
```diff
@@ -17,67 +17,159 @@
...
+	
+	// Additional control-state flags reserved for framework use.
+	TUIControlStateReserved			= 0xFF000000
 } TUIControlState;
 
+// TUIControl is the base class for control objects such as
+// buttons and sliders that convey user intent to the application.
+// You cannot use the TUIControl class directly to instantiate
+// controls. It instead defines the common interface and behavioral
+// structure for all its subclasses. The main role of UIControl is
```
Comment (jwilling) [inline lib/UIKit/TUIControl.h]: Hey now, if we're going to copy the iOS documentation we need to do it correctly, eh? **`*TUIControl`**

File: lib/UIKit/TUIControl.h [hunk 3]
```diff
@@ -17,67 +17,159 @@
...
+// events and subclasses may draw differently.
+@property (nonatomic, assign, getter = isEnabled) BOOL enabled;
+
+// For many controls, this state has no effect on behavior or
+// appearance. But other subclasses or the application object
+// might read or set this control state.
+@property (nonatomic, assign, getter = isSelected) BOOL selected;
+
+// By default, a control is not highlighted. TUIControl automatically
+// sets and clears this state automatically when a touch enters
```
Comment (jwilling) [inline lib/UIKit/TUIControl.h]: > when a *click

File: lib/UIKit/TUIControl.h [hunk 4]
```diff
@@ -98,20 +190,39 @@
 // opts to cancel it - only when the user cancels the tracking.
 - (void)endTrackingWithEvent:(NSEvent *)event;
 
-@end
-
-@interface TUIControl (TargetAction)
-
+// Add target/action for particular event. You can call this
```
Comment (jwilling) [inline lib/UIKit/TUIControl.h]: > Add a target/action for **a** particular event

File: lib/UIKit/TUIControl.h [hunk 5]
```diff
@@ -98,20 +190,39 @@
...
+// multiple times and you can specify multiple target/actions
+// for a particular event. Passing in nil as the target goes
+// up the responder chain. The action may optionally include
+// the sender and the event as parameters, in that order. The
```
Comment (jwilling) [inline lib/UIKit/TUIControl.h]: How is this dependent on the order?

File: lib/UIKit/TUIControl.h [hunk 6]
```diff
@@ -98,20 +190,39 @@
...
+// Remove the target and action for a set of events. Pass NULL
+// for the action to remove all actions for that target. You
+// may not, however, remove a block target, due to its unidentifiablity.
 - (void)removeTarget:(id)target action:(SEL)action forControlEvents:(TUIControlEvents)controlEvents;
 
-- (void)addActionForControlEvents:(TUIControlEvents)controlEvents block:(void(^)(void))action;
-
+// Get all targets, actions, and control events registered.
+// May include NSNull to indicate at least one nil target.
+// -actionsForTarget... returns NSArray of NSString selector
```
Comment (jwilling) [inline lib/UIKit/TUIControl.h]: Please put this on the line above the method itself. Also:

> Returns _an_ array of NSString _selectors_

File: lib/UIKit/TUIControl.m [hunk 7]
```diff
@@ -15,114 +15,150 @@
...
-	self = [super initWithFrame:rect];
-	if(self == nil) {
-		return nil;
+- (id)initWithFrame:(CGRect)rect {
+	if((self = [super initWithFrame:rect])) {
+		self.targetActions = [NSMutableArray array];
+		self.accessibilityTraits |= TUIAccessibilityTraitButton;
+		
+		[[NSNotificationCenter defaultCenter] addObserver:self
+												 selector:@selector(_controlTintChanged:)
```
Comment (jwilling) [inline lib/UIKit/TUIControl.m]: Why are you going through an intermediary method, `_controlTintChanged:`, rather than just using `controlTintChanged` in the first place?

File: lib/UIKit/TUIControl.m [hunk 8]
```diff
@@ -15,114 +15,150 @@
...
-	if(_controlFlags.tracking)			actual |= TUIControlStateHighlighted;
-	if(_controlFlags.highlighted)		actual |= TUIControlStateHighlighted;
-	if(![self.nsView isWindowKey])		actual |= TUIControlStateNotKey;
+	if(_controlFlags.disabled)
+		actual |= TUIControlStateDisabled;
+	if(![self.nsView isWindowKey])
+		actual |= TUIControlStateNotKey;
+	
+	if(!_controlFlags.selected) {
+		if(_controlFlags.hover)
```
Comment (jwilling) [inline lib/UIKit/TUIControl.m]: Hover isn't dependent on the control being selected.

File: lib/UIKit/TUIControl.m [hunk 9]
```diff
@@ -15,114 +15,150 @@
...
+			actual |= TUIControlStateHover;
+		
+		if(_controlFlags.tracking || _controlFlags.highlighted)
```
Comment (jwilling) [inline lib/UIKit/TUIControl.m]: This is not dependent either. The control could be selected, and the highlight would be during deselection.

File: lib/UIKit/TUIControl.m [hunk 10]
```diff
@@ -237,4 +271,103 @@ - (void)endTrackingWithEvent:(NSEvent *)event {
...
+- (void)applyStateChangeAnimated:(BOOL)animated block:(void (^)(void))block {
+	[self stateWillChange];
+	block();
+	[self stateDidChange];
+	
+	if(animated) {
+		[TUIView animateWithDuration:0.25f animations:^{
+			[self redraw];
+		}];
+	} else [self setNeedsDisplay];
```
Comment (jwilling) [inline lib/UIKit/TUIControl.m]: New line, please.

File: lib/UIKit/TUIControl.m [hunk 11]
```diff
@@ -237,4 +271,103 @@ - (void)endTrackingWithEvent:(NSEvent *)event {
...
+- (void)stateWillChange {
+	return;
+}
+
+- (void)stateDidChange {
+	return;
+}
+
+- (void)addTarget:(id)target action:(SEL)action forControlEvents:(TUIControlEvents)controlEvents {
+	if(action) {
```
Comment (jwilling) [inline lib/UIKit/TUIControl.m]: Lets do an explicit check against nil here.

File: lib/UIKit/TUIControl.m [hunk 12]
```diff
@@ -237,4 +271,103 @@ - (void)endTrackingWithEvent:(NSEvent *)event {
...
+		TUIControlTargetAction *t = [[TUIControlTargetAction alloc] init];
+		t.target = target;
+		t.action = action;
+		t.controlEvents = controlEvents;
+		[self.targetActions addObject:t];
+	}
+}
+
+- (void)addActionForControlEvents:(TUIControlEvents)controlEvents block:(void(^)(void))block {
+	if(block) {
```
Comment (jwilling) [inline lib/UIKit/TUIControl.m]: Same note as above.

File: lib/UIKit/TUIControl.m [hunk 13]
```diff
@@ -237,4 +271,103 @@ - (void)endTrackingWithEvent:(NSEvent *)event {
...
+		TUIControlTargetAction *t = [[TUIControlTargetAction alloc] init];
+		t.block = block;
+		t.controlEvents = controlEvents;
+		[self.targetActions addObject:t];
+	}
+}
+
+- (void)removeTarget:(id)target action:(SEL)action forControlEvents:(TUIControlEvents)controlEvents {
+	NSMutableArray *targetActionsToRemove = [NSMutableArray array];
+	for(TUIControlTargetAction *t in self.targetActions) {
```
Comment (jwilling) [inline lib/UIKit/TUIControl.m]: Space after `for`. Also goes for the other conditionals in this file.

File: lib/UIKit/TUIControl.m [hunk 14]
```diff
@@ -237,4 +271,103 @@ - (void)endTrackingWithEvent:(NSEvent *)event {
...
+	for(TUIControlTargetAction *t in self.targetActions)
+		e |= t.controlEvents;
+	
+	return e;
+}
+
+- (NSArray *)actionsForTarget:(id)target forControlEvent:(TUIControlEvents)controlEvent {
+	NSMutableArray *actions = [NSMutableArray array];
+	
+	for(TUIControlTargetAction *t in self.targetActions)
```
Comment (jwilling) [inline lib/UIKit/TUIControl.m]: Please add braces around this statement.

File: lib/UIKit/TUIControl.h (hunk 1 above)
Comment (avaidyam) [inline lib/UIKit/TUIControl.h]: Perfect, thanks!

File: lib/UIKit/TUIControl.h (hunk 2 above)
Comment (avaidyam) [inline lib/UIKit/TUIControl.h]: Oops! :panda_face:

File: lib/UIKit/TUIControl.h (hunk 5 above)
Comment (avaidyam) [inline lib/UIKit/TUIControl.h]: I need to make it clear that the order is as follows (for the selector): `- (void)myHappyMethod:(id)sender withEvent:(NSEvent *)event;`

File: lib/UIKit/TUIControl.h (hunk 6 above)
Comment (avaidyam) [inline lib/UIKit/TUIControl.h]: I guess I was programming too fast.... >_>

File: lib/UIKit/TUIControl.h (hunk 5 above)
Comment (jwilling) [inline lib/UIKit/TUIControl.h]: Ah you're talking about the method used _in the action_. I misread what you were trying to convey there.

File: lib/UIKit/TUIControl.m (hunk 7 above)
Comment (avaidyam) [inline lib/UIKit/TUIControl.m]: I don't think subclasses need the `NSNotification`, which simple has `NSApp` as the object and nothing more.

File: lib/UIKit/TUIControl.h (hunk 5 above)
Comment (avaidyam) [inline lib/UIKit/TUIControl.h]: It's ok, and I've made this clearer now.

File: lib/UIKit/TUIControl.m (hunk 7 above)
Comment (jwilling) [inline lib/UIKit/TUIControl.m]: What are you trying to say?

Comment (avaidyam) [inline lib/UIKit/TUIControl.m]: That it's better to abstract away the notification? :smile:

File: lib/UIKit/TUIControl.m (hunk 11 above)
Comment (avaidyam) [inline lib/UIKit/TUIControl.m]: Actually, a new trick I've discovered: `!!object` qualifies the existence of `object` through the NOT-NOT: `!!`. Mike Ash has details on his blog: http://mikeash.com/pyblog/friday-qa-2012-12-14-objective-c-pitfalls.html

File: lib/UIKit/TUIControl.m (hunk 7 above)
Comment (jwilling) [inline lib/UIKit/TUIControl.m]: I don't see how going through a two-step method call is an "abstraction". Lets say this notification was modified to call `controlTintChanged:` instead of `_ controlTintChanged:`. What difference would there be?

Comment (avaidyam) [inline lib/UIKit/TUIControl.m]: That's very true... I'll remove that "abstraction". :dash:

File: lib/UIKit/TUIControl.m (hunk 11 above)
Comment (jwilling) [inline lib/UIKit/TUIControl.m]: That is frankly horrible. Sure it works, but seriously what reason is there to use something unnecessarily obfuscated?

And that article mentioned that terrible trick as a way to convert to a boolean, which isn't what we're explicitly doing here.

**Comment by avaidyam** [comment]
@jwilling All fixes have been made. :+1:

File: lib/UIKit/TUIControl.h [hunk 15]
```diff
@@ -98,20 +189,41 @@
...
 - (void)removeTarget:(id)target action:(SEL)action forControlEvents:(TUIControlEvents)controlEvents;
 
-- (void)addActionForControlEvents:(TUIControlEvents)controlEvents block:(void(^)(void))action;
-
+// Get all targets, actions, and control events registered.
+// May include NSNull to indicate at least one nil target.
 - (NSSet *)allTargets;
 - (TUIControlEvents)allControlEvents;
+
+// -actionsForTarget... returns an NSArray of NSString selectors names or nil if none.
```
Comment (jwilling) [inline lib/UIKit/TUIControl.h]: You don't need to say `-actionsForTarget...` any more.

Comment (avaidyam) [inline lib/UIKit/TUIControl.h]: @jwilling Fixed!

File: lib/UIKit/TUIControl.m (hunk 11 above)
Comment (avaidyam) [inline lib/UIKit/TUIControl.m]: @jwilling I've done as you requested, using `!= nil`. :mailbox_with_mail:

**Comment by avaidyam** [comment]
@jwilling @CodaFi Anything else I need to fix? :mailbox_with_mail:

**Comment by avaidyam** [comment]
@jwilling :mailbox_with_mail:

File: lib/UIKit/TUIControl.h [hunk 16]
```diff
@@ -17,67 +17,158 @@
...
+// For many controls, this state has no effect on behavior or
+// appearance. But other subclasses or the application object
+// might read or set this control state.
+@property (nonatomic, assign, getter = isSelected) BOOL selected;
+
+// By default, a control is not highlighted. TUIControl automatically
+// sets and clears this state automatically when the mouse enters
+// and exits during tracking and when there is a mouse up.
+@property (nonatomic, assign, getter = isHighlighted) BOOL highlighted;
+@property (nonatomic, assign, getter = isContinuous) BOOL continuous;
```
Comment (jwilling) [inline lib/UIKit/TUIControl.h:165]: This property does not appear to be used.

File: lib/UIKit/TUIControl.m [hunk 17]
```diff
@@ -15,177 +15,208 @@
...
+@property (nonatomic, assign) TUIControlEvents controlEvents;
+
+@end
+
+@implementation TUIControlTargetAction
+@end
+
+@interface TUIControl () {
+	struct {
+		unsigned tracking:1;
```
Comment (jwilling) [inline lib/UIKit/TUIControl.m]: `unsigned int` please.

File: lib/UIKit/TUIControl.m [hunk 18]
```diff
@@ -15,177 +15,208 @@
...
 
 - (void)mouseDown:(NSEvent *)event {
-	if(_controlFlags.disabled)
+	if (_controlFlags.disabled)
 		return;
 	[super mouseDown:event];
 	
+	_controlFlags.hover = 0;
+	[self sendActionsForControlEvents:TUIControlEventMouseHoverEnded];
+	[self setNeedsDisplay];
```
Comment (jwilling) [inline lib/UIKit/TUIControl.m]: You're calling `setNeedsDisplay` here, yet you're applying a state change directly below, which means you'll redraw twice.

File: lib/UIKit/TUIControl.m [hunk 19]
```diff
@@ -15,177 +15,208 @@
...
 }
 
 - (void)mouseDown:(NSEvent *)event {
-	if(_controlFlags.disabled)
+	if (_controlFlags.disabled)
 		return;
 	[super mouseDown:event];
 	
+	_controlFlags.hover = 0;
+	[self sendActionsForControlEvents:TUIControlEventMouseHoverEnded];
```
Comment (jwilling) [inline lib/UIKit/TUIControl.m]: Why would the hover be ended? Isn't the mouse still inside the view? What the hover state really means should be clarified.

At this point it's impossible to tell from the header whether the hover state ends when the mouse goes down, or whether it ends when the mouse leaves the view, or both.

**Comment by avaidyam** [comment]
@jwilling I've fixed what you've noted :mailbox_with_mail:

File: lib/UIKit/TUIControl.m (hunk 19 above)
Comment (avaidyam) [inline lib/UIKit/TUIControl.m]: In my tests, the `mouseEntered:` and `mouseExited:` methods were being swallowed.

Comment (jwilling) [inline lib/UIKit/TUIControl.m]: Swallowed? What is that supposed to mean in this instance?

Comment (avaidyam) [inline lib/UIKit/TUIControl.m]: I don't remember exactly how I noticed it, but while adding button rollover states, the `mouseEntered:` and `mouseExited:` methods were not being called after a button press. Do you think I should just remove these segments dealing with hover (in these two methods)?

File: lib/UIKit/TUIControl.h [hunk 20]
```diff
@@ -17,68 +17,171 @@
...
+// The default method implementation does nothing, but if you
+// are subclassing a subclass of TUIControl, such as TUIButton,
+// you must call the superclass implementation of the method.
+- (void)stateWillChange;
+- (void)stateDidChange;
+
+// If the user changed the global control tint (presumably from
+// within the System Preferences application), this method will be
+// called to allow the TUIControl object to adjust itself.
+- (void)systemControlTintChanged;
```
Comment (jwilling) [inline lib/UIKit/TUIControl.h:182]: I think this would be better if we still attached the notification as it was before. This would mean reverting the block notification observer that you just added. There is some useful information contained in the user info of the notification.

Comment (avaidyam) [inline lib/UIKit/TUIControl.h:182]: Actually, nope :panda_face: Link: https://developer.apple.com/library/mac/#documentation/Cocoa/Reference/ApplicationKit/Classes/NSCell_Class/Reference/NSCell.html#//apple_ref/doc/c_ref/NSControlTintDidChangeNotification

I'm pretty sure you can get `NSApp` from anywhere linked to AppKit.

Comment (jwilling) [inline lib/UIKit/TUIControl.h:182]: Point taken. :+1:

File: lib/UIKit/TUIControl.m [hunk 21]
```diff
@@ -15,177 +15,224 @@
...
+
+- (id)initWithFrame:(CGRect)rect {
+	if ((self = [super initWithFrame:rect])) {
+		self.targetActions = [NSMutableArray array];
+		self.accessibilityTraits |= TUIAccessibilityTraitButton;
+		
+		[[NSNotificationCenter defaultCenter] addObserverForName:NSControlTintDidChangeNotification
+														  object:nil queue:nil
+													  usingBlock:^(NSNotification *note) {
+														  [self systemControlTintChanged];
```
Comment (jwilling) [inline lib/UIKit/TUIControl.m]: There's a big problem here. Can you spot it?

Comment (avaidyam) [inline lib/UIKit/TUIControl.m]: Let me take a wild guess: retain cycle. I'm obviously incorrect. =facedesk= (I'll go fix that now...)

Comment (avaidyam) [inline lib/UIKit/TUIControl.m]: Fixed it using `note.object`! Do I get brownie points for using a roundabout but concise way?

File: lib/UIKit/TUIControl.h (hunk 20 above)
Comment (avaidyam) [inline lib/UIKit/TUIControl.h:182]: I'm glad! :hamster:

File: lib/UIKit/TUIControl.m (hunk 21 above)
Comment (jwilling) [inline lib/UIKit/TUIControl.m]: Actually no, that's not a good approach. I'm actually not sure what the consequences of setting the object to self would be in this case.

But the bigger problem (besides the reference cycle) is that you're not removing the notification properly. If you read the documentation, it states that the return of this new method is:

> **Return Value**
> An opaque object to act as the observer.

So you need to keep a reference to this "token", and use that when you remove the observer in dealloc using `removeObserver:`, passing the token in.

Long story short, block-based observers aren't very nice.

Comment (avaidyam) [inline lib/UIKit/TUIControl.m]: I didn't know about that (I'm new to block-based notifications, too, if you hadn't noticed... :panda_face: ) so I'll go fix that. As to the `object` being `self`, it doesn't pose any threat according to documentation.

Comment (jwilling) [inline lib/UIKit/TUIControl.m]: To tell you the truth, the older-style observers are easier. If you must strip the notification object, do as you had done originally and pass it through to another method.

**Comment by avaidyam** [comment]
@jwilling I've fixed everything, again... :mailbox:

File: lib/UIKit/TUIControl.m (hunk 21 above)
Comment (avaidyam) [inline lib/UIKit/TUIControl.m]: This is now fixed too :]

File: lib/UIKit/TUIControl.h [hunk 22]
```diff
@@ -98,20 +206,41 @@
...
+// actions when an event occurs.
+// The first -sendAction:to:forEvent: method call is for the event
+// and is a point at which you can observe or override behavior.
+// It is then called repeately after the second call to this method.
+// To observe or modify the dispatch of action messages to targets
+// for particular events, override sendAction:to:forEvent:, evaluate
+// the passed-in selector, target object, or TUIControlEvents bit
+// mask, and proceed as required.
 - (void)sendAction:(SEL)action to:(id)target forEvent:(NSEvent *)event;
 - (void)sendActionsForControlEvents:(TUIControlEvents)controlEvents;
```
Comment (CodaFi) [inline lib/UIKit/TUIControl.h:250]: What, no documentation :penguin: ?

File: lib/UIKit/TUIControl.m [hunk 23]
```diff
@@ -237,4 +271,112 @@ - (void)endTrackingWithEvent:(NSEvent *)event {
...
+		}];
+	} else {
+		[self setNeedsDisplay];
+	}
+}
+
+// Override.
+- (void)stateWillChange {
+	return;
+}
```
Comment (CodaFi) [inline lib/UIKit/TUIControl.m:293]: I think it was cleaner to leave this without the returns.  they are implicit

File: lib/UIKit/TUIControl.m [hunk 24]
```diff
@@ -15,177 +15,215 @@
...
+													name:NSControlTintDidChangeNotification
+												  object:nil];
+}
+
+#pragma mark - Control State and Notifications
 
-- (BOOL)isEnabled
-{
-	return !_controlFlags.disabled;
+- (void)_controlTintUpdated:(NSNotification *)note {
```
Comment (jwilling) [inline lib/UIKit/TUIControl.m]: Lets try to keep our method names non-underscored. That's reserved by Apple.

File: lib/UIKit/TUIControl.h (hunk 22 above)
Comment (avaidyam) [inline lib/UIKit/TUIControl.h:250]: Alright, alright. I'll do it for the penguin. And because finals are over and I'm a second semester senior now... much time to spare! :8ball:

File: lib/UIKit/TUIControl.m (hunk 24 above)
Comment (avaidyam) [inline lib/UIKit/TUIControl.m]: Wow, that's blasphemous. Alright, I'll fix that.

**Comment by avaidyam** [comment]
@CodaFi @jwilling Done! :mailbox: Anything else?

**Comment by CodaFi** [comment]
I like it.  I'll let Mr. Willing do the honors.  /cc @jwilling

**Comment by avaidyam** [comment]
@jwilling Mr. Willing, If you'd please. :)

File: lib/UIKit/TUIControl.h [hunk 25]
```diff
@@ -17,68 +17,176 @@
...
 	
-	TUIControlEventMouseHover			= 1 <<  9,
+	// A mouse hover begin event that a control is tracking.
+	TUIControlEventMouseHoverBegan		= 1 <<  9,
+	
+	// A mouse hover end event that a control stops tracking.
+	TUIControlEventMouseHoverEnded		= 1 <<  10,
+	
+	// A manipulated control caused to emit a series of different values.
 	TUIControlEventValueChanged			= 1 << 12,
```
Comment (jwilling) [inline lib/UIKit/TUIControl.h]: Why was `12` chosen? I know it was there before, but is this because of UIKit?

File: lib/UIKit/TUIControl.h [hunk 26]
```diff
@@ -17,68 +17,176 @@
 #import "TUIView.h"
 
 typedef enum TUIControlEvents : NSUInteger {
+	
+	// A mouse down event in the control.
 	TUIControlEventMouseDown			= 1 <<  0,
```
Comment (jwilling) [inline lib/UIKit/TUIControl.h]: This bit mask is either using two spaces or a tab after the `<<`, whereas below they use a single space. Please correct this. Same goes for some below as well.

File: lib/UIKit/TUIControl.h (hunk 25 above)
Comment (avaidyam) [inline lib/UIKit/TUIControl.h]: Yep, because of UIKit. Odd, but I didn't call the shots and changing it now would potentially break many things (bad coders...?)

File: lib/UIKit/TUIControl.h (hunk 26 above)
Comment (avaidyam) [inline lib/UIKit/TUIControl.h]: Fixed.

File: lib/UIKit/TUIControl.m [hunk 27]
```diff
@@ -234,7 +268,115 @@ - (BOOL)continueTrackingWithEvent:(NSEvent *)event {
 }
 
 - (void)endTrackingWithEvent:(NSEvent *)event {
+	// Unimplemented.
```
Comment (jwilling) [inline lib/UIKit/TUIControl.m]: Saying "unimplemented" here could lead a person to believe that this is incomplete. If you want to leave a comment in, please make it more clear. For example:

> Implemented by subclasses

Comment (avaidyam) [inline lib/UIKit/TUIControl.m]: Fixed that too! :]

**Comment by avaidyam** [comment]
Any other little things left to catch?

**Comment by jwilling** [comment]
Thanks!

**Comment by avaidyam** [comment]
@jwilling Can the button PR be reopened now?
